packages/meteor/setimmediate.js

```diff
@@ -1,141 +1,4 @@
-
-function usePostMessage() {
-  // The test against `importScripts` prevents this implementation
-  // from being installed inside a web worker, where
-  // `global.postMessage` means something completely different and
-  // can't be used for this purpose.
-
-  if (!global.postMessage || global.importScripts) {
-    return null;
-  }
-
-  // Avoid synchronous post message implementations.
-
-  var postMessageIsAsynchronous = true;
-  var oldOnMessage = global.onmessage;
-  global.onmessage = function () {
-      postMessageIsAsynchronous = false;
-  };
-  global.postMessage("", "*");
-  global.onmessage = oldOnMessage;
-
-  if (! postMessageIsAsynchronous)
-    return null;
-
-  var funcIndex = 0;
-  var funcs = {};
-
-  // Installs an event handler on `global` for the `message` event: see
-  // * https://developer.mozilla.org/en/DOM/window.postMessage
-  // * http://www.whatwg.org/specs/web-apps/current-work/multipage/comms.html#crossDocumentMessages
-
-  // XXX use Random.id() here?
-  var MESSAGE_PREFIX = "Meteor._setImmediate." + Math.random() + '.';
-
-  function isStringAndStartsWith(string, putativeStart) {
-    return (typeof string === "string" &&
-            string.substring(0, putativeStart.length) === putativeStart);
-  }
-
-  function onGlobalMessage(event) {
-    // This will catch all incoming messages (even from other
-    // windows!), so we need to try reasonably hard to avoid letting
-    // anyone else trick us into firing off. We test the origin is
-    // still this window, and that a (randomly generated)
-    // unpredictable identifying prefix is present.
-    if (event.source === global &&
-        isStringAndStartsWith(event.data, MESSAGE_PREFIX)) {
-      var index = event.data.substring(MESSAGE_PREFIX.length);
-      try {
-        if (funcs[index])
-          funcs[index]();
-      }
-      finally {
-        delete funcs[index];
-      }
-    }
-  }
-
-  if (global.addEventListener) {
-    global.addEventListener("message", onGlobalMessage, false);
-  } else {
-    global.attachEvent("onmessage", onGlobalMessage);
-  }
-
-  var setImmediate = function (fn) {
-    // Make `global` post a message to itself with the handle and
-    // identifying prefix, thus asynchronously invoking our
-    // onGlobalMessage listener above.
-    ++funcIndex;
-    funcs[funcIndex] = fn;
-    global.postMessage(MESSAGE_PREFIX + funcIndex, "*");
-  };
-  setImmediate.implementation = 'postMessage';
-  return setImmediate;
-}
-
-
-function useTimeout() {
-  var setImmediate = function (fn) {
-    global.setTimeout(fn, 0);
-  };
-  setImmediate.implementation = 'setTimeout';
-  return setImmediate;
-}
-
-
-Meteor._setImmediate =
-  useSetImmediate() ||
-  usePostMessage() ||
-  useTimeout();
+/**
+ * Make sure that `setImmediate` works both on the server and the client. Uses a microtask in the client.
+ */
+Meteor._setImmediate = Meteor.isServer ? setImmediate : fn => Promise.resolve().then(fn)
```

Comment:
setImmediate needs to schedule a macrotask. The purpose of the api is to schedule some work after any already queued macrotasks, without adding unnecessary delays like setTimeout would (which is why the browser polyfills usually use postMessage). Using a microtask would do the opposite - schedule the work to run before any other macrotask.

For example, I use it in one app to break apart a large amount of synchronous work so it finishes as quickly as possible without making the page unresponsive. Microtasks would instead block the browser from running event handlers or rendering until it finished.

This also will break how the DDP queueing works on the client with method stubs since it relies on setImmediate creating macrotasks.

Reply:
Oh, I see. Will adjust, thanks!

Reply:
I adjusted and included a comment there with this info.